src/utils/graph.ts

```diff
@@ -65,10 +65,14 @@ export const addEdge = (edgeParams: Edge | Connection, elements: Elements): Elem
       throw new Error(`Can't create edge. Node with id=${id} does not exist.`);
     }
   });
-  // need to check if handles exist
-  // 
-  // 
-  // 
+  
+  // make sure that the handles exists in each node
+  const handleElements = Array.from(document.getElementsByClassName("react-flow__handle")) as HTMLDivElement[]
```

Comment:
We are storing some informations about the handles in the node's `__rf` object. Couldn't we use this in order to check the handles here?

Reply:
I also thought that was the case, but upon further investigation `__rf` is not always available. `addEdge` is passed elements which can have nodes in it. But the nodes are not necessarily the node objects provided by the store. For example, in Validation, `addEdge` is passed the state array `elements` not the store array `elements`. And `__rf` is only available from nodes provided by the store. At least that's my understanding.

Reply:
good point ;) But can't we check wether the node id exists (in the passed element array) or the node with the corresponding handleId (if one is passed via `sourceHandle ` or `targetHandle`).

Reply:
I'm sorry I'm not following. The nodeID existence check is happening in the baseline code already. And there's not enough information being passed in by `elements` to check if the handleID exists within the nodeIDs specified. 

If there is an assumption that `addEdge` will only ever be called as a callback function by `onConnect` I guess there's no need to do the check because you're guaranteed that handleID exist within the nodeIDs specified.

Reply:
Ok, I think we can remove this check. `addEdge` should only be used as a helper for the `onConnect` handler.